### PR TITLE
use new octet-stream API for read() on SPO driver

### LINKED ISSUE #3999: Move SPO driver to new blob API
SPO has added a new API for "data" blobs, which can be used for both "snapshot" and "data" blobs. Currently only the "data" blob methods use this new API, but for the sake of simplicity, we should move everything over to the new API.

From @vladsud:
>For SPO, we should use new API to read both "snapshot" and "image" blobs (old "snapshot" blob API does not support new blobs)
> - That said, there is difference in payload. Old API was returning json, and thus base64 encoded payload, vs. new API returns binary. We should be careful here in SPO driver - for now encode it as base64, but eventually change the model for driver to return (there is an issue on that somewhere assigned to me).
> - As a staging, we can expose an extra API from driver to allow both reading binary & base64 output, such that we do not waist resources where consumer and producer requirements are already matched

### PR DESCRIPTION
**Pull request by wes-carlson**
Fix #3999

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 1]
```diff
@@ -225,28 +225,32 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
...
+    private async readWithEncodingOutput(blobId: string, outputFormat: string): Promise<string> {
+        // const blob = this.blobCache.get(blobId);
+        let blob: IBlob | undefined;
+        assert(!blob || blob.encoding === "base64" || blob.encoding === undefined, "wrong blob encoding format");
+        let buffer;
         // Reset the timer on attempted cache read
         this.scheduleClearBlobsCache();
         if (blob === undefined) {
-            this.checkSnapshotUrl();
+            this.checkAttachmentGETUrl();
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:247]: Any chance we can start removing these kinds of APIs (where possible) in favor of getters that do same function - throw if uri is not present?

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 2]
```diff
@@ -255,45 +259,41 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
                         headers: Object.keys(headers).length !== 0 ? true : undefined,
                         waitQueueLength: this.epochTracker.rateLimiter.waitQueueLength,
                     },
-                    async () => this.epochTracker.fetchAndParseAsJSON<IBlob>(url, { headers }, FetchType.blob),
+                    async () => this.epochTracker.fetchResponse(url, { headers }, FetchType.blob),
                 );
             });
-            blob = response.content;
+            buffer = await response.arrayBuffer();
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:264]: @ChumpChief - I'm a bit confused here, maybe I missed something in your PR. I see we call  this.scheduleClearBlobsCache() at the top, but we never cache this blob. Given that we do not cache, should we not reset timer?

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 3]
```diff
@@ -255,45 +258,41 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
...
+                if (outputFormat === blob.encoding || (outputFormat === "string" && blob.encoding === undefined))  {
+                    return blob.content;
+                } else if (outputFormat === "base64") {
+                    return fromUtf8ToBase64(blob.content);
+                } else {
+                    return fromBase64ToUtf8(blob.content);
+                }
             }
+
+            return IsoBuffer.from(buffer).toString(outputFormat === "base64" ? "base64" : "utf8");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:278]: This does not look right. transformation of attributes content should happen in the same way, no matter if it blob came from cache or not.

Comment (wes-carlson) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:278]: Fixing the branch name is done below, not here (unless I'm misunderstanding you)

Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:278]: Conversion to UTF8 does not look right.

[RFC 4627 ](https://tools.ietf.org/html/rfc4627)(see Chapter 3 - Encoding) that governs application/json content type says that payload can be in any 5 formats (utf-8/16/32, BE or LE). When SPO receives such summary payload, I'd guess they will get proper in-memory representation (likely UTF-16, given that's default for most environments on Windows), and will likely store it in that format. Thus when we read data, it's likely to give us data in this format. They can move to Linux boxes in the future and accidently get UTF-32 (default everywhere else except Windows).

I guess the key - current protocol does not describe what encoding they store data in, and if that has material impact on this API. We can change it (and ensure it's part of protocol in some form or shape). If not, this code should auto detect encoding in the same way as RFC 4627 instructs browsers (and other entities) to treat Unicode formatting

Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:278]: That all said, I think we are already in this world before your changes. Receiving "base64" input (as we get today) does not say how original buffer was generated, i.e. was it utf-8 or utf-16 before encoding it with base64. Our (runtime) utilities use utf-8, but when we send application/json payload, and when server processes input, it loses encoding (or rather does not provide us it). So I guess it's already utf-8 all the way through. I'll talk to SPO folks about it

**Review by vladsud — APPROVED** [review]
I think it's good to go (assuming we tested all interesting combinations of how we write data, and how we read data) - having some stand-alone tests for that would be great!
I've started thread with SPO folks on UTF encoding on their side to understand better if we need auto-detection logic on client, or we put in writing that utf-8 is the encoding used by server.

**Pull request merged**